platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/InputChangesContext.java

```diff
@@ -21,11 +21,11 @@
 import javax.annotation.Nullable;
 import java.util.Optional;
 
-public class InputChangesContext extends ValidationFinishedContext {
+public class InputChangesContext extends CachingContext {
 
     private final InputChangesInternal inputChanges;
 
-    public InputChangesContext(ValidationFinishedContext parent, @Nullable InputChangesInternal inputChanges) {
+    public InputChangesContext(CachingContext parent, @Nullable InputChangesInternal inputChanges) {
```

Comment:
❓ Why is this change necessary?

Reply:
If I don't do this, the chaining doesn't work any more in ExecutionServices. More precisely, then context were not a strict hierarchy, and one of the two hierarchies did not include `CachingContext`: 
<img width="514" alt="image" src="https://github.com/gradle/gradle/assets/423186/c29bd2e0-ae24-4ad0-aa95-e7b4542e667f">

I was actually surprised that `InputChangesContext` isn't a `CachingContext`.

Reply:
I remember that being intentional. I wanted to go towards a more independent, less linear-looking context hierarchy. But it never worked. I think it's fine to change it the way you did. We should come back to this with generated context implementations sometime later.